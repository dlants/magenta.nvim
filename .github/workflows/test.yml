name: Tests

on:
  push:
    branches: [main]
  pull_request:
    branches: [main]

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3

      - name: Setup Node.js
        uses: actions/setup-node@v3
        with:
          node-version: "18"
          cache: "npm"

      - name: Install dependencies
        run: npm install --frozen-lockfile

      - name: Setup Neovim
        uses: rhysd/action-setup-vim@v1
        with:
          neovim: true
          version: stable

      - name: Install typescript-language-server
        run: npm install -g typescript-language-server typescript

      - name: Run tests
        run: npx vitest

  lint-and-format:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        command:
          - name: "typecheck"
            run: "npx tsc --noEmit"
          - name: "eslint"
            run: "npx eslint ."
          - name: "prettier"
            run: "npx prettier --check ."
    name: ${{ matrix.command.name }}
    steps:
      - uses: actions/checkout@v3

      - name: Setup Node.js
        uses: actions/setup-node@v3
        with:
          node-version: "18"
          cache: "npm"

      - name: Install dependencies
        run: npm install --frozen-lockfile

      - name: Run ${{ matrix.command.name }}
        run: ${{ matrix.command.run }}
