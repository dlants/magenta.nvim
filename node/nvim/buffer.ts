import type { Nvim } from "./nvim-node";
import type { Position0Indexed, Position1Indexed, Row0Indexed } from "./window";
import { withTimeout } from "../utils/async";
import type { AbsFilePath, UnresolvedFilePath } from "../utils/files";
import type { ExtmarkId, ExtmarkOptions } from "./extmarks";

export type Line = string & { __line: true };
export type BufNr = number & { __bufnr: true };
export type Mode = "n" | "i" | "v";

/**
 * Branded type for Neovim namespace IDs.
 */
export type NamespaceId = number & { __namespaceId: true };

/**
 * Well-known namespace for magenta highlighting system.
 * This ensures all magenta highlights are grouped together and can be cleared as a unit.
 */
export const MAGENTA_HIGHLIGHT_NAMESPACE = "magenta-highlights";

export class NvimBuffer {
  constructor(
    public readonly id: BufNr,
    private nvim: Nvim,
  ) {}

  getOption(option: string) {
    return this.nvim.call("nvim_buf_get_option", [this.id, option]);
  }

  setOption(option: string, value: unknown) {
    return this.nvim.call("nvim_buf_set_option", [this.id, option, value]);
  }

  getChangeTick() {
    return this.nvim.call("nvim_buf_get_changedtick", [
      this.id,
    ]) as unknown as Promise<number>;
  }

  setLines({
    start,
    end,
    lines,
  }: {
    start: Row0Indexed;
    end: Row0Indexed;
    lines: Line[];
  }) {
    return this.nvim.call("nvim_buf_set_lines", [
      this.id,
      start,
      end,
      false,
      lines,
    ]);
  }

  async getLines({
    start,
    end,
  }: {
    start: Row0Indexed;
    end: Row0Indexed;
  }): Promise<Line[]> {
    // Ensure buffer is loaded before getting lines
    // unloaded buffers return no lines, see https://github.com/neovim/neovim/pull/8660
    await this.nvim.call("nvim_eval", [`bufload(${this.id})`]);

    const lines = await this.nvim.call("nvim_buf_get_lines", [
      this.id,
      start,
      end,
      false,
    ]);
    return lines as Line[];
  }

  async getText({
    startPos,
    endPos,
  }: {
    startPos: Position0Indexed;
    endPos: Position0Indexed;
  }): Promise<Line[]> {
    const lines = await this.nvim.call("nvim_buf_get_text", [
      this.id,
      startPos.row,
      startPos.col,
      endPos.row,
      endPos.col,
      {},
    ]);
    return lines as Line[];
  }

  setText({
    startPos,
    endPos,
    lines,
  }: {
    startPos: Position0Indexed;
    endPos: Position0Indexed;
    lines: Line[];
  }): Promise<void> {
    return this.nvim.call("nvim_buf_set_text", [
      this.id,
      startPos.row,
      startPos.col,
      endPos.row,
      endPos.col,
      lines,
    ]);
  }

  setMark({ mark, pos }: { mark: string; pos: Position1Indexed }) {
    return this.nvim.call("nvim_buf_set_mark", [
      this.id,
      mark,
      pos.row,
      pos.col,
      {},
    ]);
  }

  setSiderbarKeymaps() {
    return this.nvim.call("nvim_exec_lua", [
      `require("magenta.keymaps").set_sidebar_buffer_keymaps(${this.id})`,
      [],
    ]);
  }

  setDisplayKeymaps() {
    return this.nvim.call("nvim_exec_lua", [
      `require("magenta.keymaps").set_display_buffer_keymaps(${this.id})`,
      [],
    ]);
  }

  setInlineKeymaps(targetBufnr: BufNr) {
    return this.nvim.call("nvim_exec_lua", [
      `require("magenta.keymaps").set_inline_buffer_keymaps(${this.id}, ${targetBufnr})`,
      [],
    ]);
  }

  getName(): Promise<UnresolvedFilePath> {
    return this.nvim.call("nvim_buf_get_name", [
      this.id,
    ]) as Promise<UnresolvedFilePath>;
  }

  setName(name: string) {
    return this.nvim.call("nvim_buf_set_name", [this.id, name]);
  }

  async attemptWrite() {
    // TODO: this is really gross. Unfortunately when write fails in this context, for
    // some reason we never hear back from this promise.
    // This is brittle, since if the file takes longer than the timeout to write, we may
    // incorrectly assume that it failed to write. However, this should be rare. The extra
    // 1s delay is also not ideal, but we should be doing this rarely - only when responding
    // to a tool call. As such it should be ok for now, and I really just want to move on from
    // this issue.
    // See https://github.com/neovim/neovim/discussions/33804 for further discussion
    return withTimeout(
      this.nvim.call("nvim_exec_lua", [
        `\
vim.api.nvim_buf_call(${this.id}, function()
  -- silent to avoid blocking on unable to write (which would typically alert the user)
  vim.cmd("silent! write")
end)`,
        [],
      ]),
      1000,
    );
  }

  async attemptEdit() {
    // TODO: this is really gross. Unfortunately when write fails in this context, for
    // some reason we never hear back from this promise.
    // This is brittle, since if the file takes longer than the timeout to write, we may
    // incorrectly assume that it failed to write. However, this should be rare. The extra
    // 1s delay is also not ideal, but we should be doing this rarely - only when responding
    // to a tool call. As such it should be ok for now, and I really just want to move on from
    // this issue.
    // See https://github.com/neovim/neovim/discussions/33804 for further discussion
    return withTimeout(
      this.nvim.call("nvim_exec_lua", [
        `\
vim.api.nvim_buf_call(${this.id}, function()
  -- silent to avoid blocking on unable to write (which would typically alert the user)
  vim.cmd("silent! edit")
end)`,
        [],
      ]),
      1000,
    );
  }

  static async create(listed: boolean, scratch: boolean, nvim: Nvim) {
    const bufNr = (await nvim.call("nvim_create_buf", [
      listed,
      scratch,
    ])) as BufNr;
    return new NvimBuffer(bufNr, nvim);
  }

  delete(options?: { force?: boolean; unload?: boolean }) {
    return this.nvim.call("nvim_buf_delete", [this.id, options || {}]);
  }

  isValid(): Promise<boolean> {
    return this.nvim.call("nvim_buf_is_valid", [this.id]);
  }

  static async bufadd(absolutePath: AbsFilePath, nvim: Nvim) {
    const bufNr = (await nvim.call("nvim_eval", [
      `bufadd("${absolutePath}")`,
    ])) as BufNr;
    await nvim.call("nvim_eval", [`bufload(${bufNr})`]);
    return new NvimBuffer(bufNr, nvim);
  }

  // Extmark methods

  /**
   * Set an extmark in this buffer with the given options.
   * Returns the extmark ID for later updates or deletion.
   */
  async setExtmark({
    startPos,
    endPos,
    options,
  }: {
    startPos: Position0Indexed;
    endPos: Position0Indexed;
    options: ExtmarkOptions;
  }): Promise<ExtmarkId> {
    const namespaceId = await this.getMagentaNamespace();

    // Prepare extmark options with end position
    const extmarkOpts = {
      ...options,
      end_row: endPos.row,
      end_col: endPos.col,
    };

    const extmarkId = await this.nvim.call("nvim_buf_set_extmark", [
      this.id,
      namespaceId,
      startPos.row,
      startPos.col,
      extmarkOpts,
    ]);

    return extmarkId as ExtmarkId;
  }

  /**
   * Delete a specific extmark from this buffer.
   */
  async deleteExtmark(extmarkId: ExtmarkId): Promise<void> {
    const namespaceId = await this.getMagentaNamespace();
    await this.nvim.call("nvim_buf_del_extmark", [
      this.id,
      namespaceId,
      extmarkId,
    ]);
  }

  /**
   * Clear all extmarks in the magenta highlight namespace for this buffer.
   * This is useful for bulk cleanup when unmounting views or clearing highlights.
   */
  async clearAllExtmarks(): Promise<void> {
    const namespaceId = await this.getMagentaNamespace();

    // Clear all extmarks in the namespace for this buffer
    await this.nvim.call("nvim_buf_clear_namespace", [
      this.id,
      namespaceId,
      0, // start line
      -1, // end line (-1 means end of buffer)
    ]);
  }

  /**
   * Update an existing extmark with new options and/or position.
   * This is more efficient than deleting and recreating for position/style changes.
   */
  async updateExtmark({
    extmarkId,
    startPos,
    endPos,
    options,
  }: {
    extmarkId: ExtmarkId;
    startPos: Position0Indexed;
    endPos: Position0Indexed;
    options: ExtmarkOptions;
  }): Promise<ExtmarkId> {
    const namespaceId = await this.getMagentaNamespace();

    // Prepare extmark options with end position and existing ID
    const extmarkOpts = {
      ...options,
      id: extmarkId,
      end_row: endPos.row,
      end_col: endPos.col,
    };

    const updatedId = await this.nvim.call("nvim_buf_set_extmark", [
      this.id,
      namespaceId,
      startPos.row,
      startPos.col,
      extmarkOpts,
    ]);

    return updatedId as ExtmarkId;
  }

  /**
   * Get all extmarks in the magenta namespace for this buffer.
   * Returns an array of extmark information including ID, position, and options.
   */
  async getExtmarks(): Promise<
    Array<{
      id: ExtmarkId;
      startPos: Position0Indexed;
      endPos: Position0Indexed;
      options: ExtmarkOptions;
    }>
  > {
    const namespaceId = await this.getMagentaNamespace();

    // Get all extmarks in the namespace
    const extmarks = await this.nvim.call("nvim_buf_get_extmarks", [
      this.id,
      namespaceId,
      0, // start position
      -1, // end position (-1 means end of buffer)
      { details: true }, // include details like end position and options
    ]);

    return (extmarks as unknown[][]).map((extmarkData) =>
      this.parseExtmarkData(extmarkData),
    );
  }

  /**
   * Get a specific extmark by its ID from the magenta namespace.
   * Returns undefined if the extmark doesn't exist.
   */
  async getExtmarkById(extmarkId: ExtmarkId): Promise<
    | {
        id: ExtmarkId;
        startPos: Position0Indexed;
        endPos: Position0Indexed;
        options: ExtmarkOptions;
      }
    | undefined
  > {
    const namespaceId = await this.getMagentaNamespace();

    try {
      // Get the specific extmark by ID
      const extmarksResult = await this.nvim.call("nvim_buf_get_extmarks", [
        this.id,
        namespaceId,
        extmarkId, // start from this specific extmark ID
        extmarkId, // end at this specific extmark ID
        { details: true, limit: 1 }, // include details and limit to 1 result
      ]);

      const extmarksArray = extmarksResult as unknown[][];
      if (extmarksArray.length === 0) {
        return undefined;
      }

      return this.parseExtmarkData(extmarksArray[0]);
    } catch {
      // If the extmark doesn't exist, nvim_buf_get_extmarks may throw
      return undefined;
    }
  }

  /**
   * Parse raw extmark data from nvim_buf_get_extmarks into our structured format.
   */
  private parseExtmarkData(extmarkData: unknown[]): {
    id: ExtmarkId;
    startPos: Position0Indexed;
    endPos: Position0Indexed;
    options: ExtmarkOptions;
  } {
    const [id, startRow, startCol, details] = extmarkData;
    return {
      id: id as ExtmarkId,
      startPos: { row: startRow, col: startCol } as Position0Indexed,
      endPos: {
        row: (details as { end_row: unknown }).end_row || startRow,
        col: (details as { end_col: unknown }).end_col || startCol,
      } as Position0Indexed,
      options: details as ExtmarkOptions,
    };
  }

  /**
   * Create or get the magenta highlighting namespace.
   * Uses a well-known namespace name for consistency across views.
   */
  async getMagentaNamespace(): Promise<NamespaceId> {
    const namespaceId = await this.nvim.call("nvim_create_namespace", [
      MAGENTA_HIGHLIGHT_NAMESPACE,
    ]);
    return namespaceId as NamespaceId;
  }
}
