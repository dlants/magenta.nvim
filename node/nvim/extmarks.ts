// Core highlight types and interfaces for the magenta.nvim VDOM system

/**
 * Union type of all available highlight groups for type safety.
 * Includes standard Neovim groups, Treesitter groups, LSP semantic tokens, and dynamic groups.
 */
/**
 * All available highlight groups for type safety.
 * Includes standard Neovim groups, Treesitter groups, and LSP semantic tokens.
 */
export const HL_GROUPS = [
  // Standard Neovim highlight groups
  "ErrorMsg",
  "WarningMsg",
  "Directory",
  "String",
  "Comment",
  "Bold",
  "Identifier",
  "Underlined",
  "Normal",
  "Function",
  // Treesitter groups
  "@variable",
  "@function.call",
  "@keyword",
  "@string",
  "@comment",
  "@type",
  "@constant",
  // LSP semantic token groups
  "@lsp.type.variable",
  "@lsp.type.function",
  "@lsp.type.keyword",
  "@lsp.type.string",
  "@lsp.type.comment",
  "@lsp.type.type",
  "@lsp.type.constant",
] as const;

/**
 * Union type of all available highlight groups for type safety.
 */
export type HLGroup = (typeof HL_GROUPS)[number];

/**
 * Text styling options for bold, italic, underline, etc.
 */
export type TextStyle = {
  bold?: boolean;
  italic?: boolean;
  underline?: boolean;
  strikethrough?: boolean;
};

/**
 * Custom color styling when semantic groups aren't sufficient.
 * Use sparingly to maintain colorscheme compatibility.
 */
export type ColorStyle = {
  fg?: string;
  bg?: string;
};

/**
 * Branded type for tracking extmark IDs from Neovim.
 */
export type ExtmarkId = number & { __extmarkId: true };

/**
 * Comprehensive options for nvim_buf_set_extmark.
 * Covers all available extmark functionality.
 */
export type ExtmarkOptions = {
  // Basic highlighting
  hl_group?: HLGroup | TextStyleGroup;
  hl_eol?: boolean;
  hl_mode?: "replace" | "combine" | "blend";
  priority?: number;

  // Line-level styling
  line_hl_group?: HLGroup;

  // Sign column
  sign_text?: string;
  sign_hl_group?: HLGroup;

  // Line number styling
  number_hl_group?: HLGroup;

  // Advanced features
  conceal?: string;
  url?: string;

  // Position control
  right_gravity?: boolean;
  end_right_gravity?: boolean;
  strict?: boolean;

  // Persistence
  undo_restore?: boolean;
  invalidate?: boolean;
  ephemeral?: boolean;
};

export type TextStyleGroup = string & { __textStyleGroup: true };

/**
 * Create a text style highlight group dynamically.
 * Returns highlight group name for use in ExtmarkOptions.
 * Note: This creates dynamic highlight groups that may need to be registered with Neovim.
 */
export function createTextStyleGroup(style: TextStyle): TextStyleGroup {
  const parts: string[] = [];
  if (style.bold) parts.push("bold");
  if (style.italic) parts.push("italic");
  if (style.underline) parts.push("underline");
  if (style.strikethrough) parts.push("strikethrough");
  return parts.join(",") as TextStyleGroup;
}

/**
 * Compare two ExtmarkOptions objects for equality.
 * Used for efficient diffing during updates to avoid unnecessary extmark operations.
 */
export function extmarkOptionsEqual(
  options1: ExtmarkOptions | undefined,
  options2: ExtmarkOptions | undefined,
): boolean {
  // Both undefined - equal
  if (options1 === undefined && options2 === undefined) {
    return true;
  }

  // One undefined, one defined - not equal
  if (options1 === undefined || options2 === undefined) {
    return false;
  }

  // Compare all relevant properties
  return (
    options1.hl_group === options2.hl_group &&
    options1.hl_eol === options2.hl_eol &&
    options1.hl_mode === options2.hl_mode &&
    options1.priority === options2.priority &&
    options1.line_hl_group === options2.line_hl_group &&
    options1.sign_text === options2.sign_text &&
    options1.sign_hl_group === options2.sign_hl_group &&
    options1.number_hl_group === options2.number_hl_group &&
    options1.conceal === options2.conceal &&
    options1.url === options2.url &&
    options1.right_gravity === options2.right_gravity &&
    options1.end_right_gravity === options2.end_right_gravity &&
    options1.strict === options2.strict &&
    options1.undo_restore === options2.undo_restore &&
    options1.invalidate === options2.invalidate &&
    options1.ephemeral === options2.ephemeral
  );
}
