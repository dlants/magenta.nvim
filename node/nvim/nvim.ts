import type { Nvim } from "./nvim-node";
import { NvimBuffer, type BufNr } from "./buffer.ts";
import {
  NvimWindow,
  type Position1IndexedCol1Indexed,
  type WindowId,
} from "./window.ts";
import type { NvimCwd } from "../utils/files.ts";

export function getOption(option: string, nvim: Nvim) {
  return nvim.call("nvim_get_option", [option]);
}

export async function getCurrentWindow(nvim: Nvim) {
  const winId = (await nvim.call("nvim_get_current_win", [])) as WindowId;
  return new NvimWindow(winId, nvim);
}

export async function getCurrentBuffer(nvim: Nvim) {
  const bufnr = (await nvim.call("nvim_get_current_buf", [])) as BufNr;
  return new NvimBuffer(bufnr, nvim);
}

export async function getAllBuffers(nvim: Nvim) {
  const bufs = (await nvim.call("nvim_list_bufs", [])) as BufNr[];
  return bufs.map((bufnr) => new NvimBuffer(bufnr, nvim));
}

export async function getAllWindows(nvim: Nvim) {
  const winIds = (await nvim.call("nvim_list_wins", [])) as WindowId[];
  return winIds.map((winId) => new NvimWindow(winId, nvim));
}

export async function getcwd(nvim: Nvim): Promise<NvimCwd> {
  const res = await nvim.call("nvim_exec2", [
    "echo getcwd()",
    { output: true },
  ]);
  if (typeof res.output != "string") {
    throw new Error(`Unable to get cwd`);
  }
  return res.output as NvimCwd;
}

export async function mode(nvim: Nvim) {
  const res = await nvim.call("nvim_exec2", ["echo mode()", { output: true }]);
  if (typeof res.output != "string") {
    throw new Error(`Unable to get mode`);
  }
  return res.output;
}

export async function getpos(
  nvim: Nvim,
  str: string,
): Promise<Position1IndexedCol1Indexed> {
  const res = await nvim.call("nvim_exec2", [
    `echo getpos("${str}")`,
    { output: true },
  ]);
  if (typeof res.output != "string") {
    throw new Error(`Unable to getpos`);
  }
  const posStr = res.output;

  const [_, row, col] = posStr
    // eslint-disable-next-line no-useless-escape
    .replace(/[\[\]]/g, "")
    .split(",")
    .map((n) => parseInt(n.trim(), 10));

  return {
    row,
    col,
  } as Position1IndexedCol1Indexed;
}

export function diffthis(nvim: Nvim) {
  return nvim.call("nvim_command", ["diffthis"]);
}

export interface QuickfixEntry {
  bufnr: number;
  lnum: number;
  col: number;
  text: string;
  type?: string;
  filename?: string;
  valid?: number;
}

export async function getQuickfixList(nvim: Nvim): Promise<QuickfixEntry[]> {
  const qflist = await nvim.call("nvim_exec2", [
    "echo json_encode(getqflist())",
    { output: true },
  ]);
  if (typeof qflist.output !== "string") {
    throw new Error("Unable to get quickfix list");
  }
  return JSON.parse(qflist.output) as QuickfixEntry[];
}

export async function quickfixListToString(
  entries: QuickfixEntry[],
  nvim: Nvim,
): Promise<string> {
  const lines: string[] = [];

  for (const entry of entries) {
    let filename: string;
    if (entry.filename) {
      filename = entry.filename;
    } else if (entry.bufnr > 0) {
      // Get the filename from buffer number
      const bufname = await nvim.call("nvim_buf_get_name", [entry.bufnr]);
      filename = bufname ? bufname : `buffer ${entry.bufnr}`;
    } else {
      filename = `buffer ${entry.bufnr}`;
    }

    const line = entry.lnum > 0 ? `:${entry.lnum}` : "";
    const col = entry.col > 0 ? `:${entry.col}` : "";
    lines.push(`${filename}${line}${col}: ${entry.text}`);
  }

  return lines.join("\n");
}

export async function notify(nvim: Nvim, message: string) {
  const luaScript = `
      vim.notify(
        [[${message}]],
        vim.log.levels.INFO
      )
    `;

  return nvim.call("nvim_exec_lua", [luaScript, []]);
}

export function notifyErr(nvim: Nvim, err: Error | string, ...rest: unknown[]) {
  return nvim.call("nvim_notify", [
    `Unexpected error:
${err instanceof Error ? err.message : JSON.stringify(err)}
${err instanceof Error ? err.stack : ""}
${rest.length ? JSON.stringify(rest) : ""}`,
    3,
    {},
  ]);
}
