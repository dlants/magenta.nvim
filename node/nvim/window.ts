import type { Nvim } from "./nvim-node";
import { NvimBuffer, type BufNr } from "./buffer.ts";

export type Row0Indexed = number & { __row0Indexed: true };
export type Row1Indexed = number & { __row1Indexed: true };
export type ByteIdx = number & { __byteIdx: true };
export type ByteIdx1Indexed = number & { __byteIdx1Indexed: true };

/** A coordinate in a js string, which are utf-16 encoded by default. This is the coordinate that lsp clients typically expect.
 */
export type StringIdx = number & { __charIdx: true };

export type PositionString = {
  row: Row0Indexed;
  col: StringIdx;
};

export type Position1Indexed = {
  row: Row1Indexed;
  col: ByteIdx;
};

export type Position1IndexedCol1Indexed = {
  row: Row1Indexed;
  col: ByteIdx1Indexed;
};

export type Position0Indexed = {
  row: Row0Indexed;
  col: ByteIdx;
};

export function pos0to1(pos: Position0Indexed): Position1Indexed {
  return {
    row: pos.row + 1,
    col: pos.col,
  } as Position1Indexed;
}

export function pos1to0(pos: Position1Indexed): Position0Indexed {
  return {
    row: pos.row - 1,
    col: pos.col,
  } as Position0Indexed;
}

export function pos1col1to0(
  pos: Position1IndexedCol1Indexed,
): Position0Indexed {
  return {
    row: pos.row - 1,
    col: Math.max(0, pos.col - 1),
  } as Position0Indexed;
}

export type WindowId = number & { __winId: true };

export class NvimWindow {
  constructor(
    public readonly id: WindowId,
    private nvim: Nvim,
  ) {}

  valid(): Promise<boolean> {
    return this.nvim.call("nvim_win_is_valid", [this.id]);
  }

  clearjumps() {
    return this.nvim.call("nvim_command", [
      `call win_execute(${this.id}, 'clearjumps')`,
    ]);
  }

  setWidth(width: number) {
    return this.nvim.call("nvim_win_set_width", [this.id, width]);
  }

  getOption(name: string) {
    return this.nvim.call("nvim_win_get_option", [this.id, name]);
  }

  setOption(name: string, value: unknown) {
    return this.nvim.call("nvim_win_set_option", [this.id, name, value]);
  }

  async getVar(name: string) {
    try {
      return await this.nvim.call("nvim_win_get_var", [this.id, name]);
    } catch (e) {
      this.nvim.logger.warn(`getVar(${name}) failed: ${JSON.stringify(e)}`);
      return undefined;
    }
  }

  setVar(name: string, value: unknown) {
    return this.nvim.call("nvim_win_set_var", [this.id, name, value]);
  }

  close(force: boolean = false) {
    return this.nvim.call("nvim_win_close", [this.id, force]);
  }

  async buffer(): Promise<NvimBuffer> {
    const bufNr = (await this.nvim.call("nvim_win_get_buf", [
      this.id,
    ])) as BufNr;
    return new NvimBuffer(bufNr, this.nvim);
  }

  async getCursor(): Promise<Position1Indexed> {
    const [row, col] = await this.nvim.call("nvim_win_get_cursor", [this.id]);
    return { row, col } as Position1Indexed;
  }

  setCursor(pos: Position1Indexed) {
    return this.nvim.call("nvim_win_set_cursor", [this.id, [pos.row, pos.col]]);
  }

  zt() {
    return this.nvim.call("nvim_exec2", [
      `call win_execute(${this.id}, 'normal! zt')`,
      {},
    ]);
  }

  zb() {
    return this.nvim.call("nvim_exec2", [
      `call win_execute(${this.id}, 'normal! zb')`,
      {},
    ]);
  }

  /**
   * Get the window position (row, column) in the grid
   * @returns Promise<[row, col]> where (0,0) is top-left
   */
  async getPosition(): Promise<[number, number]> {
    return await this.nvim.call("nvim_win_get_position", [this.id]);
  }

  async topLine(): Promise<number> {
    const res = await this.nvim.call("nvim_exec2", [
      `echo line('w0', "${this.id}")`,
      { output: true },
    ]);

    return Number(res.output);
  }
}
