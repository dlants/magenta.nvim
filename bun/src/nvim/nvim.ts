import { context } from "../context.ts";
import { NvimBuffer, type BufNr } from "./buffer.ts";
import { NvimWindow, type WindowId } from "./window.ts";

export function getOption(option: string) {
  return context.nvim.call("nvim_get_option", [option]);
}

export async function getCurrentWindow() {
  const winId = (await context.nvim.call(
    "nvim_get_current_win",
    [],
  )) as WindowId;
  return new NvimWindow(winId);
}

export async function getAllBuffers() {
  const bufs = (await context.nvim.call("nvim_list_bufs", [])) as BufNr[];
  return bufs.map((bufnr) => new NvimBuffer(bufnr));
}

export async function getAllWindows() {
  const winIds = (await context.nvim.call("nvim_list_wins", [])) as WindowId[];
  return winIds.map((winId) => new NvimWindow(winId));
}

export async function getcwd() {
  const res = await context.nvim.call("nvim_exec2", [
    "echo getcwd()",
    { output: true },
  ]);
  if (typeof res.output != "string") {
    throw new Error(`Unable to get cwd`);
  }
  return res.output;
}

export function diffthis() {
  return context.nvim.call("nvim_command", ["diffthis"]);
}
