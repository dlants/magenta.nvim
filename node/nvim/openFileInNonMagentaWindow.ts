import type { Nvim } from "./nvim-node";
import { getAllWindows, getcwd } from "../nvim/nvim";
import { NvimBuffer } from "../nvim/buffer";
import type { WindowId } from "../nvim/window";
import type { MagentaOptions } from "../options";
import {
  resolveFilePath,
  type AbsFilePath,
  type UnresolvedFilePath,
} from "../utils/files";

export async function openFileInNonMagentaWindow(
  filePath: UnresolvedFilePath | AbsFilePath,
  context: { nvim: Nvim; options: MagentaOptions },
): Promise<void> {
  try {
    const windows = await getAllWindows(context.nvim);
    const nonMagentaWindows = [];
    const magentaWindows = [];

    // Find all non-magenta windows and magenta windows
    for (const window of windows) {
      const isMagenta = await window.getVar("magenta");
      if (!isMagenta) {
        nonMagentaWindows.push(window);
      } else {
        magentaWindows.push(window);
      }
    }

    let targetWindowId: WindowId | null = null;

    // if there are non-magenta windows, use one of those
    if (nonMagentaWindows.length > 0) {
      targetWindowId = nonMagentaWindows[0].id;
    }

    const absFilePath = resolveFilePath(await getcwd(context.nvim), filePath);
    // Open the buffer in the target window or create a new window if needed
    const fileBuffer = await NvimBuffer.bufadd(absFilePath, context.nvim);

    if (targetWindowId) {
      // Open in the existing window
      await context.nvim.call("nvim_win_set_buf", [
        targetWindowId,
        fileBuffer.id,
      ]);
    } else if (nonMagentaWindows.length === 0 && magentaWindows.length > 0) {
      // Use the configured sidebarPosition from options
      const sidebarPosition = context.options.sidebarPosition;
      // Open on the opposite side
      const newWindowSide = sidebarPosition === "left" ? "right" : "left";

      // Open a new window on the appropriate side
      await context.nvim.call("nvim_open_win", [
        fileBuffer.id,
        true, // Enter the window
        {
          win: -1, // Global split
          split: newWindowSide,
        },
      ]);
    } else {
      // No suitable window found, create a new one
      await context.nvim.call("nvim_command", [`split ${absFilePath}`]);
    }
  } catch (error) {
    context.nvim.logger?.error(
      `Error opening file ${filePath}: ${(error as Error).message}`,
    );
  }
}
