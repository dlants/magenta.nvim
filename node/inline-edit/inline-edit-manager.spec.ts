import { describe, expect, it } from "vitest";
import { withDriver } from "../test/preamble";
import type { ToolRequestId } from "../tools/toolManager";
import { getCurrentBuffer, getCurrentWindow } from "../nvim/nvim";
import type { Line } from "../nvim/buffer";
import type { Position0Indexed } from "../nvim/window";

describe("node/inline-edit/inline-edit-manager.spec.ts", () => {
  it("performs inline edit on file", async () => {
    await withDriver(async (driver) => {
      await driver.editFile("node/test/fixtures/poem.txt");
      const targetBuffer = await getCurrentBuffer(driver.nvim);
      await driver.startInlineEdit();

      // Verify inline edit window opened
      await driver.assertWindowCount(2);

      const inputWindow = await getCurrentWindow(driver.nvim);
      const winbar = await inputWindow.getOption("winbar");
      expect(winbar).toEqual("Magenta Inline Prompt");

      const mode = await driver.nvim.call("nvim_get_mode", []);
      expect(mode).toEqual({ mode: "i", blocking: false });

      const inputBuffer = await getCurrentBuffer(driver.nvim);
      await inputBuffer.setLines({
        start: 0,
        end: -1,
        lines: ["Please change 'Silver' to 'Golden' in line 2"] as Line[],
      });
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      driver.submitInlineEdit(targetBuffer.id);
      const request = await driver.mockAnthropic.awaitPendingInlineRequest();
      expect(request.messages).toMatchSnapshot();

      const modifiable = await inputBuffer.getOption("modifiable");
      expect(modifiable).toBe(false);

      const inputLines = await inputBuffer.getLines({ start: 0, end: -1 });
      expect(inputLines.join("\n")).toEqual("Input sent, awaiting response...");

      await driver.mockAnthropic.respondInline({
        stopReason: "end_turn",
        inlineEdit: {
          status: "ok",
          value: {
            id: "id" as ToolRequestId,
            name: "inline-edit",
            input: {
              find: "Silver shadows dance with ease.",
              replace: "Golden shadows dance with ease.",
            },
          },
        },
      });

      await driver.assertBufferContains(
        inputBuffer,
        `\
Got tool use: `,
      );

      await driver.assertBufferContains(
        targetBuffer,
        `\
>>>>>>> Suggested change
Golden shadows dance with ease.
=======
Silver shadows dance with ease.
<<<<<<< Current`,
      );
    });
  });

  it("can do multiple inline edits on same file", async () => {
    await withDriver(async (driver) => {
      await driver.editFile("node/test/fixtures/poem.txt");
      const targetBuffer = await getCurrentBuffer(driver.nvim);
      await driver.startInlineEdit();
      await driver.assertWindowCount(2);

      {
        const inputWindow = await getCurrentWindow(driver.nvim);
        await inputWindow.close();
      }
      await driver.assertWindowCount(1);

      // open inline edit again
      await driver.startInlineEdit();
      await driver.assertWindowCount(2);

      const inputBuffer = await getCurrentBuffer(driver.nvim);
      await inputBuffer.setLines({
        start: 0,
        end: -1,
        lines: ["Please change 'Silver' to 'Golden' in line 2"] as Line[],
      });
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      driver.submitInlineEdit(targetBuffer.id);
      await driver.mockAnthropic.awaitPendingInlineRequest();

      const inputLines = await inputBuffer.getLines({ start: 0, end: -1 });
      expect(inputLines.join("\n")).toEqual("Input sent, awaiting response...");
    });
  });

  it("performs inline edit with selection", async () => {
    await withDriver(async (driver) => {
      await driver.editFile("node/test/fixtures/poem.txt");
      const targetBuffer = await getCurrentBuffer(driver.nvim);

      // Select a range of text
      await driver.selectRange(
        { row: 1, col: 0 } as Position0Indexed,
        { row: 1, col: 32 } as Position0Indexed,
      );

      await driver.startInlineEditWithSelection();
      await driver.assertWindowCount(2);

      const inputBuffer = await getCurrentBuffer(driver.nvim);
      await inputBuffer.setLines({
        start: 0,
        end: -1,
        lines: ["Please change 'Silver' to 'Golden'"] as Line[],
      });

      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      driver.submitInlineEdit(targetBuffer.id);
      const request = await driver.mockAnthropic.awaitPendingReplaceRequest();
      expect(request.messages).toMatchSnapshot();
    });
  });
});
