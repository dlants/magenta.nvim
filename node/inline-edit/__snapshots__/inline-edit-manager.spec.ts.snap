// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`node/inline-edit/inline-edit-manager.spec.ts > performs inline edit on file 1`] = `
[
  {
    "content": "I am working in file \`node/test/fixtures/poem.txt\` with the following contents:
\`\`\`txt
Moonlight whispers through the trees,
Silver shadows dance with ease.
Stars above like diamonds bright,
Paint their stories in the night.
\`\`\`

Please change 'Silver' to 'Golden' in line 2",
    "role": "user",
  },
]
`;
